### PR TITLE
Use arange and repeat for deterministic bincount

### LINKED ISSUE #1413: Validation with torchmetrics extremely slow
### Bug description

Hi all,

I recently tried to implement a DeepLabV3 training pipeline. I wanted to use the build-in `torchmetrics.JaccardIndex` as my evaluation metric. My `LightningModule`  looks like this:

```python
import torchmetrics 
from pytorch_lightning import LightningModule
from torchvision.models.segmentation.deeplabv3 import deeplabv3_resnet50


class DeepLabV3LightningModule(LightningModule):
    def __init__(self):
        self.model = deeplabv3_resnet50(
            num_classes=38,
            aux_loss=False
        )
        self.loss = nn.CrossEntropyLoss(ignore_index=255, reduction="mean")
        self.iou_metric = torchmetrics.JaccardIndex(
            task="multiclass", 
            threshold=0.5, 
            num_classes=38,
            average="macro",
        )

    def training_step(self, batch, batch_idx):
        imgs, masks = batch
        out = self.model(imgs)
        preds = out["out"]
        loss = self.loss(preds, masks)       
        return loss

    def validation_step(self, batch, batch_idx):
        imgs, masks = batch
        out = self.model(imgs)
        preds = out["out"]
        loss = self.loss(preds, masks)
        preds = torch.softmax(preds, dim=1)
        pred_labels = torch.argmax(preds, dim=1)
        
        # measure runtime of metric update
        start = timer()
        self.iou_metric.update(pred_labels, masks)
        elapsed = timer() - start
        return elapsed

    def validation_epoch_end(self, outputs):
        avg_runtime = round(mean(outputs), 4)
        print(f"GPU {self.local_rank}: {avg_runtime} seconds")

```

When using this validation procedure, it is extremely slow. On average, the update step of the metric takes 23.4 seconds. However, the first 3 updates are very fast (<1 second), then they become slow.

I tried to reproduce this behavior in a MWE:

```python
from timeit import default_timer as timer
from statistics import mean
import torchmetrics
import torch

num_classes = 38

iou_metric = torchmetrics.JaccardIndex(
    task="multiclass",
    threshold=0.5, 
    num_classes=num_classes,
    average="macro"
).to("cuda")

# dummy labels in shape [b, h, w]
label_mask = torch.randint(low=0, high=num_classes-1, size=(8, 480, 640), device="cuda")

# dummy predicted labels in shape [b, h, w]
pred_mask = torch.randint(low=0, high=num_classes-1, size=(8, 480, 640), device="cuda")


runtime_hist = []
for i in range(100):
    start = timer()
    iou_metric.update(label_mask, pred_mask)
    elapsed = timer() - start
    runtime_hist.append(elapsed)


avg_runtime = round(mean(runtime_hist), 2)
print(avg_runtime)
```

Here I get an average update duration of 0.03 seconds, so I do not encounter the extremely slow update as in my `LightningModule` above. To me this looks like there is something wrong. At this point, I am not sure if thi

Here some training information for my pytorch-lightning training pipeline:
- OS: Ubuntu 20.04.4
- CUDA 11.3
- DDP training strategy
- GPUs: 4x V100
- batch size: 8
- image size (width x height): 640 x 480
- number of workers in dataloader: 8

My package versions:
- pytorch lightning: 1.8.4.post0 (installed via pip)
- torch: 1.13.0
- torchvision: 0.14.0
- torchmetrics: 0.11.0
- numpy: 11.23.5

Thanks so much!
Lukas

### How to reproduce the bug

_No response_

### Error messages and logs

_No response_

### Environment

_No response_

### More info

_No response_

cc @carmocca @justusschock @awaelchli @borda

### PR DESCRIPTION
**Pull request by kyle-dorman**
## What does this PR do?

Fixes #1413

Uses torch.arange and  torch.repeat instead of a for loop to deterministically calculate bincount. Should run MUCH faster now. 

<details>
  <summary>Before submitting</summary>

- [ Y] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [Y ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [Y ] Did you make sure to **update the docs**?
- [N/A ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2184.org.readthedocs.build/en/2184/

### CODE REVIEW

File: src/torchmetrics/utilities/data.py [hunk 1]
```diff
@@ -191,11 +191,12 @@ def _bincount(x: Tensor, minlength: Optional[int] = None) -> Tensor:
...
     if minlength is None:
         minlength = len(torch.unique(x))
+
     if torch.are_deterministic_algorithms_enabled() or _XLA_AVAILABLE or _TORCH_GREATER_EQUAL_1_12 and x.is_mps:
-        output = torch.zeros(minlength, device=x.device, dtype=torch.long)
-        for i in range(minlength):
-            output[i] = (x == i).sum()
-        return output
+        size = len(x)
+        mesh = torch.meshgrid(torch.arange(size), torch.arange(minlength), indexing="ij")[1].to(x.device)
```
Comment (SkafteNicki) [inline src/torchmetrics/utilities/data.py:197]:
```suggestion
        mesh = torch.arange(minlength, device=x.device).repeat(size, 1)
```
some small experimentation tells me this should be faster
image

**Review by Borda — APPROVED** [review]

**Comment by daturkel** [comment]
Is there a blocker for merging this change? My team would love to incorporate it as the speed of validation metrics is currently slowing down our ability to iterate.

**Comment by SkafteNicki** [comment]
Hi @daturkel, sorry nothing in principal stopping this from being merged, we have just not been around for the last couple of weeks. Trying to pick up the work now.
There is a problem with our CI that is currently blocking all PRs. When that is fixed, this should hopefully get merged.

**Comment by daturkel** [comment]
> Hi @daturkel, sorry nothing in principal stopping this from being merged, we have just not been around for the last couple of weeks. Trying to pick up the work now. There is a problem with our CI that is currently blocking all PRs. When that is fixed, this should hopefully get merged.

All good @SkafteNicki , I appreciate the team's effort! Thanks for the update

File: src/torchmetrics/utilities/data.py [hunk 2]
```diff
@@ -174,7 +174,7 @@ def _bincount(x: Tensor, minlength: Optional[int] = None) -> Tensor:
         - deterministic mode on GPU.
         - MPS devices
 
-    This implementation fallback to a for-loop counting occurrences in that case.
+    The implementation uses torch.arange for deterministic counting only.
```
Comment (awaelchli) [inline src/torchmetrics/utilities/data.py:177]: The word "only" at the end is problematic I think, because the code says otherwise. It also uses the slower implementation in case of MPS and XLA

**Review by awaelchli — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]
